api/src/main/java/org/openmrs/module/openconceptlab/importer/Importer.java

```diff
@@ -97,8 +97,70 @@ public void setSaver(Saver persister) {
 	    this.saver = persister;
     }
 
-	public void setZipFile(ZipFile zipFile) {
-		this.zipFile = zipFile;
+	public void setImportFile(File importFile) {
+		this.importFile = importFile;
+	}
+
+	/**
+	 * This will return an appropriate input stream, depending on whether the passed file is a Zip or Json text file
+	 */
+	private void openInputStream() throws IOException {
+		if (importFile != null) {
+			if (importFile.getName().endsWith(".zip")) {
+				ZipFile zipFile = new ZipFile(importFile);
+				in = new CountingInputStream(Utils.extractExportInputStreamFromZip(zipFile));
+			} else if (importFile.getName().endsWith(".json")) {
+				in = new CountingInputStream(FileUtils.openInputStream(importFile));
+			}
+			else {
+				throw new IllegalStateException("Import file " + importFile.getName() + " must be either a zip or json file");
+			}
+			String subscriptionUrl = importFile.getAbsolutePath();
+			importService.updateSubscriptionUrl(anImport, subscriptionUrl);
+		}
+		else {
+			Subscription subscription = importService.getSubscription();
+			Import lastImport = importService.getLastSuccessfulSubscriptionImport();
+			Date updatedSince = null;
+			if (lastImport != null) {
+				updatedSince = lastImport.getOclDateStarted();
+			}
+
+			OclResponse oclResponse;
+
+			if (updatedSince == null) {
+				oclResponse = oclClient.fetchOclConcepts(subscription.getUrl(), subscription.getToken());
+				String releaseVersion = oclClient.getOclReleaseVersion(subscription.getUrl(), subscription.getToken());
+				importService.updateReleaseVersion(anImport, releaseVersion);
+			}
+			else {
+				if (subscription.isSubscribedToSnapshot()) {
+					oclResponse = oclClient.fetchSnapshotUpdates(subscription.getUrl(), subscription.getToken(), updatedSince);
+				}
+				else {
+					oclResponse = oclClient.fetchOclConcepts(subscription.getUrl(), subscription.getToken(), lastImport.getReleaseVersion());
+					importService.updateReleaseVersion(anImport,
+							oclClient.getOclReleaseVersion(subscription.getUrl(), subscription.getToken()));
+				}
+			}
+
+			if (oclResponse != null) {
+				importService.updateOclDateStarted(anImport, oclResponse.getUpdatedTo());
+				importService.updateSubscriptionUrl(anImport, subscription.getUrl());
+				in = new CountingInputStream(oclResponse.getContentStream());
+				totalBytesToProcess = oclResponse.getContentLength();
+			}
+		}
+	}
+
+	private void closeInputStream() {
+		IOUtils.closeQuietly(in);
+	}

```

Comment:
I'll need to look back at the code but there was a reason I needed to do it this way.  I think it was because there are 2 paths here - one in which an Input Stream is coming from a File, and another in which an Input Stream is coming from the subscription.  If we can make this work with a try-with-resources, I'm all for it, but we'd need to look at why this is like this.